src/search/pdbs/pattern_collection_generator_multiple_cegar.cc

```diff
@@ -0,0 +1,300 @@
+#include "pattern_collection_generator_multiple_cegar.h"
+
+#include "cegar.h"
+#include "pattern_database.h"
+#include "utils.h"
+
+#include "../option_parser.h"
+#include "../plugin.h"
+
+#include "../utils/countdown_timer.h"
+#include "../utils/markup.h"
+#include "../utils/rng.h"
+#include "../utils/rng_options.h"
+
+#include <vector>
+
+using namespace std;
+
+namespace pdbs {
+PatternCollectionGeneratorMultipleCegar::PatternCollectionGeneratorMultipleCegar(
+    options::Options &opts)
+    : max_pdb_size(opts.get<int>("max_pdb_size")),
+      max_collection_size(opts.get<int>("max_collection_size")),
+      wildcard_plans(opts.get<bool>("wildcard_plans")),
+      cegar_max_time(opts.get<double>("max_time")),
+      verbosity(opts.get<utils::Verbosity>("verbosity")),
+      rng(utils::parse_rng_from_options(opts)),
+      random_seed(opts.get<int>("random_seed")),
+      stagnation_limit(opts.get<double>("stagnation_limit")),
+      blacklist_trigger_percentage(opts.get<double>("blacklist_trigger_percentage")),
+      blacklist_on_stagnation(opts.get<bool>("blacklist_on_stagnation")),
+      total_max_time(opts.get<double>("total_max_time")) {
+}
+
+PatternCollectionInformation PatternCollectionGeneratorMultipleCegar::generate(
+    const shared_ptr<AbstractTask> &task) {
+    if (verbosity >= utils::Verbosity::NORMAL) {
+        utils::g_log << "Generating patterns using the multiple CEGAR algorithm" << endl;
+    }
+
+    TaskProxy task_proxy(*task);
+    utils::CountdownTimer timer(total_max_time);
+
+    // Store the set of goals in random order.
+    vector<FactPair> goals = get_goals_in_random_order(task_proxy, rng);
+
+    // Store the non-goal variables for potential blacklisting.
+    int num_vars = task_proxy.get_variables().size();
+    vector<int> non_goal_variables;
+    non_goal_variables.reserve(num_vars - goals.size());
+    for (int var_id = 0; var_id < num_vars; ++var_id) {
+        bool is_goal_var = false;
+        for (const FactPair &goal : goals) {
+            if (var_id == goal.var) {
+                is_goal_var = true;
+                break;
+            }
+        }
+        if (!is_goal_var) {
+            non_goal_variables.push_back(var_id);
+        }
+    }
+
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "goal variables: ";
+        for (auto goal : goals) {
+            utils::g_log << goal.var << ", ";
+        }
+        utils::g_log << endl;
+        utils::g_log << "non-goal variables: " << non_goal_variables << endl;
+    }
+
+    // Collect all unique patterns and their PDBs.
+    utils::HashSet<Pattern> generated_patterns;
+    shared_ptr<PDBCollection> generated_pdbs = make_shared<PDBCollection>();
+
+    bool blacklisting = false;
+    double stagnation_start_time = -1;
+    int num_iterations = 1;
+    int goal_index = 0;
+    int remaining_collection_size = max_collection_size;
+    const utils::Verbosity cegar_verbosity(utils::Verbosity::SILENT);
```

Comment:
Is there a reason why a different verbosity is passed to CEGAR?

Reply:
Yes. Even with normal verbosity, there would be way too much output. Also, I don't think that the multiple CEGAR algorithm should care about how the pattern it gets was generated. It produces output itself whenever it think this is useful, such as when turning blacklisting on. Maybe it could produce even more output on the way, as to have some visual kind of progress. But dumping something every iteration is just too verbose.